112. Path Sum/PathSumStep2.java

```diff
@@ -0,0 +1,160 @@
+            if (node == null) {
+                return false;
+            }
+            rest -= node.val;
+            if (node.left == null && node.right == null) {
+                return rest == 0;
+            }
+            return hasPathSumHelper(node.left, rest) || hasPathSumHelper(node.right, rest);
+        }
+    }
+    /*
+     * 思ったこと
+     * ・葉ノードであること、値は負にもなることを盛り込めば書ける。
+     */
+    
+    
+    // DFS
+    class Solution2_2 {
+        public boolean hasPathSum(TreeNode root, int targetSum) {
+            if (root == null) {
+                return false;
+            }
+            targetSum -= root.val;
+            if (root.left == null && root.right == null) {
+                return targetSum == 0;
+            }
+            if (hasPathSum(root.left, targetSum)) {
+                return true;
+            }
+            return hasPathSum(root.right, targetSum);
+        }
+    }
+    /*
+     * 思ったこと
+     * ・1本の道を見つけたらtrueでいいか。
+     */
+    
+    
+    // DFS
+    class Solution2_3 {
+        record NodeFrame(TreeNode node, int rest) {}
+        public boolean hasPathSum(TreeNode root, int targetSum) {
+            Stack<NodeFrame> nodeFrames = new Stack<>();
+            nodeFrames.push(new NodeFrame(root, targetSum));
+            while(!nodeFrames.isEmpty()) {
+                NodeFrame nodeFrame = nodeFrames.pop();
+                TreeNode node = nodeFrame.node();
+                int rest = nodeFrame.rest();
+                if (node == null) {
+                    continue;
+                }
+                rest -= node.val;
+                if (node.left == null && node.right == null) {
+                    if (rest == 0) {
+                        return true;
+                    }
+                    continue;
+                }
+                nodeFrames.push(new NodeFrame(node.left, rest));
+                nodeFrames.push(new NodeFrame(node.right, rest));
+            }
+            return false;
+        }
+    }
+    /*
+     * 思ったこと
+     * ・はじめ、sumを足していく方式にしようとしたが、rootがnullの時にwhileでまとめて処理できなかった。
+     * と思ったけど、0を設定しておけばよかった。
+     */
+    // DFS 2_3の足していく方式
+    class Solution2_4 {
+        record NodeFrame(TreeNode node, int sum) {}
+        public boolean hasPathSum(TreeNode root, int targetSum) {
+            Stack<NodeFrame> nodeFrames = new Stack<>();
+            nodeFrames.push(new NodeFrame(root, 0));
+            while (!nodeFrames.isEmpty()) {
+                NodeFrame nodeFrame = nodeFrames.pop();
+                TreeNode node = nodeFrame.node();
+                int sum = nodeFrame.sum();
+                if (node == null) {
+                    continue;
+                }
+                sum += node.val;
+                if (node.left == null && node.right == null) {
+                    if (sum == targetSum) {
+                        return true;
+                    }
+                    continue;
+                }

```

Comment:
```suggestion
                if (node.left == null && node.right == null && sum == targetSum) {
                        return true;
                }
```
nestを減らせそうです。どうせnullは上で弾いているので

Reply:
ありがとうございます。
確かにどっちみち null なので弾けますが、流れを追い終える時点が若干先延ばしになると思います。
大げさかもしれませんが、プログラムの処理をきれいにするために認知負荷を上げているような気がします。
continue すれば、葉ノードだからそりゃその先は見なくていいよね、で読みやすいんじゃないかなと思っています。

Reply:
あとnullをnodeFramesに突っ込まなくて済みますね。